### PR TITLE
feat(plugins) allow cross-namespace plugin references

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:

Add a new colon-delimited value format for the `konghq.com/plugins` CSV annotation. The string before the colon indicates a remote KongPlugin namespace. This namespace must contain a ReferenceGrant that allows this reference.

While the existing `konghq.com/plugins: foo,bar` format loads plugins `foo` and `bar` from the same namespace as the annotated resource, the new `konghq.com/plugins: foo,alpha:bar` format loads `foo` from the local namespace and `bar` from the `alpha` namespace.

[example_cross_plugin.yaml.txt](https://github.com/Kong/kubernetes-ingress-controller/files/15242556/example_cross_plugin.yaml.txt) provides a more complete example. Service `one` in namespace `pwuh` loads the `test` KongPlugin from the `pwuh` namespace as usual. Service `two` in namespace `fwuh` loads the same `test` KongPlugin from `pwuh`. It can do so because a ReferenceGrant in `pwuh` grants KongPlugin access to `fwuh` Services.

**Which issue this PR fixes**:

Fix https://github.com/Kong/kubernetes-ingress-controller/issues/5355. Although the original request concerned KongClusterPlugins, using KongPlugins allows for a simpler security configuration model.

Review https://github.com/Kong/kubernetes-ingress-controller/issues/5355#issuecomment-2075580376 for additional background on why multi-entity plugins require grants to adhere to the Kubernetes namespace security model.

**Special notes for your reviewer**:

There was already a fair amount of code duplication in the plugin relationship builder. Ideally we'd do away with it, but I've left it around for the time being. The `isRemotePluginReferenceAllowed()` helper hopefully negates the worst of it by extracting most of the logic we need to duplicate into a separate function, so the duplicated code is just the function invocation.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/annotations/annotations.go [hunk 1]
```diff
@@ -155,6 +155,36 @@ func ExtractKongPluginsFromAnnotations(anns map[string]string) []string {
 	return kongPluginCRs
 }
 
+type NamespacedKongPlugin struct {
+	Namespace string
+	Name      string
+}
```
Comment (pmalek) [inline internal/annotations/annotations.go:161]: This could potentially be a type alias for https://pkg.go.dev/k8s.io/apimachinery/pkg/types#NamespacedName.

File: internal/annotations/annotations.go [hunk 2]
```diff
@@ -155,6 +155,36 @@ func ExtractKongPluginsFromAnnotations(anns map[string]string) []string {
...
+
+// ExtractNamespacedKongPluginsFromAnnotations extracts a KongPlugin name and optional namespace from an annotation
+// value. Plugins are delimited by ",". Values are either colon-delimited "namespace:name" strings or name-only
+// strings.
+func ExtractNamespacedKongPluginsFromAnnotations(anns map[string]string) []NamespacedKongPlugin {
```
Comment (pmalek) [inline internal/annotations/annotations.go:166]: Can we have a UT for this?

File: internal/annotations/annotations.go [hunk 3]
```diff
@@ -155,6 +155,36 @@ func ExtractKongPluginsFromAnnotations(anns map[string]string) []string {
...
+	var plugins []NamespacedKongPlugin
+	v := pluginsFromAnnotations(anns)
+	if v == "" {
+		return plugins
+	}
+	for _, s := range strings.Split(v, ",") {
```
Comment (pmalek) [inline internal/annotations/annotations.go:172]: `plugins` could preallocate `len(strings.Split(v, ","))` entries

```suggestion
	v := pluginsFromAnnotations(anns)
	if v == "" {
		return nil
	}
	split := strings.Split(v, ",")
	plugins := make([]NamespacedKongPlugin, 0, len(split))

	for _, s := range split {
```

File: internal/gatewayapi/contraints.go [hunk 4]
```diff
@@ -18,5 +18,13 @@ type RouteT interface {
 
 type BackendRefT interface {
 	BackendRef |
-		SecretObjectReference
+		SecretObjectReference |
+		// TODO TRR the plugin isn't really a backend in this sense, so that's a bit messy
+		// Do we need to distinguish between backend and other types of reference? at what level?
```
Comment (pmalek) [inline internal/gatewayapi/contraints.go:23]: My memory might fail me but as far as I recall this was only here to contrain the allowed types here.

Given that this only needs `Namespace`, `Name`, `Kind` and `Group` in `IsRefAllowedByGrant` you could remove all those type and replace the type constraint with `client.Object`. Do we want that? 🤷 It's up to us at what level we want to constrain the API.

File: internal/gatewayapi/references.go
Comment (pmalek) [inline internal/gatewayapi/references.go:1]: Nit: moving this in a separate PR would make the review a bit more focused

File: internal/dataplane/kongstate/kongstate.go [hunk 5]
```diff
@@ -330,7 +335,32 @@ func (ks *KongState) FillVaults(
...
+			if reflect.DeepEqual(from, otherFrom) {
+				allowed[gatewayapi.Namespace(grant.ObjectMeta.Namespace)] = append(allowed[gatewayapi.Namespace(grant.ObjectMeta.Namespace)], grant.Spec.To...)
+			}
+		}
+	}
+
+	return allowed
+}
+
+func (ks *KongState) getPluginRelations(cacheStore store.Storer, log logr.Logger) map[string]util.ForeignRelations {
```
Comment (pmalek) [inline internal/dataplane/kongstate/kongstate.go:363]: This `func` got a bit lengthy with this change, didn't it? I would suggest we consider refactoring this (a tiny bit) so that at the very least the added code that iterate over `ks.Services` and `ks.Consumers` is extracted out. WDYT?

File: internal/gatewayapi/references.go
Comment (rainest) [reply internal/gatewayapi/references.go:1]: Continued in another thread.

File: internal/gatewayapi/contraints.go (hunk 4 above)
Comment (rainest) [reply internal/gatewayapi/contraints.go:23]: Alright, so basically no reason to not make it more or less generic for anything with a GVK and NamespacedName AFAICT. Re

> Nit: moving this in a separate PR would make the review a bit more focused

in `internal/gatewayapi/references.go`: at present this is mostly just moving what was previously in `internal/dataplane/translator/wrappers_refchecker.go` (since it's now used across packages, and would have caused an import cycle at its old location) aside the new case for plugins. How do you feel about a TODO issue to push that refactor to a separate PR?

If I _don't_ do that refactor it's just a relocation and a case addition in this PR, with the expectation that the cases will all collapse after refactoring.

File: internal/dataplane/kongstate/kongstate.go (hunk 5 above)
Comment (rainest) [reply internal/dataplane/kongstate/kongstate.go:363]: This is the

> Ideally we'd do away with it

from OP. I played around with:

- Extracted the new logic to a separate function that accepts a `type addPluginRelationFunc func(namespace, plugin, identifier string)` arg. Having a standalone helper that accepts a dynamically-instantiated function bound to some dynamic scope `map[string]util.ForeignRelations` instance felt icky.
- Making `map[string]util.ForeignRelations` a defined type with methods to add relations. This also resulted in a mess.

So IDK exactly how I'd want to refactor it, but I don't think there's a simple way to. Was there a smaller change you were thinking of? We're already iterating over everything that can bind to plugins, but the additions mean those iterated tasks are much more complicated, and unfortunately not in a way that actually separates out cleanly AFAICT.

File: internal/gatewayapi/contraints.go (hunk 4 above)
Comment (pmalek) [reply internal/gatewayapi/contraints.go:23]: I have noticed that. My suggestion was just to make the PR more targeted so that the reviews are more focused, PRs smaller and then potential/hypothetical backports require less effort.

I'm OK with leaving this as is.

File: internal/dataplane/kongstate/kongstate.go (hunk 5 above)
Comment (pmalek) [reply internal/dataplane/kongstate/kongstate.go:363]: Something like this could work (although it could still be improved):

https://github.com/Kong/kubernetes-ingress-controller/compare/feat/plugin-across...refactor-getPluginRelations?expand=1

https://github.com/Kong/kubernetes-ingress-controller/commit/c53cd3c5bfd1eb2e6241a21c6697351c3e5e713c

**Comment by rainest** [comment]
Some sort of nonsense apparently means the test... passes? but then leaves some of its resources around, furthermore preventing the test suite from finishing :(

File: internal/dataplane/kongstate/kongstate.go [hunk 6]
```diff
@@ -330,81 +334,147 @@ func (ks *KongState) FillVaults(
...
 
+type pluginReference struct {
+	Referer   client.Object
+	Namespace string
+	Name      string
+}
+
+func isRemotePluginReferenceAllowed(s store.Storer, r pluginReference) error {
+	// remote plugin. considered part of this namespace if a suitable ReferenceGrant exists
+	grants, err := s.ListReferenceGrants()
```
Comment (randmonkey) [inline internal/dataplane/kongstate/kongstate.go:449]: Are there any performance concerns to run `ListReferenceGrants()` each time when a resource (`Service`/`Consumer`/...) refers to remote plugins? If we believe that there will not be significant amount of cross-namespace reference, this may be OK.

File: internal/gatewayapi/references.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+		return isRefAllowedByGrant(
+			(*string)(br.Namespace),
+			(string)(br.Name),
+			(string)(*br.Group),
+			(string)(*br.Kind),
+			allowedRefs,
+		)
+
+	case PluginLabelReference:
```
Comment (randmonkey) [inline internal/gatewayapi/references.go:79]: Could we add UTs for the `SecretObjectReference` and `PluginLabelReference` cases?

**Review by randmonkey — COMMENTED** [review]
Mostly LGTM, some minor comments remaining.

File: internal/dataplane/kongstate/kongstate.go (hunk 6 above)
Comment (rainest) [reply internal/dataplane/kongstate/kongstate.go:449]: I think it's mostly an upstream concern--we don't do much other than piggyback off the https://pkg.go.dev/k8s.io/client-go/tools/cache#Store `List()` implementations, and basically trust that the Kubernetes core devs make those calls efficient because they're used for everything, and would provide additional filters or special cases if it became a problem (similar to https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#motivation).

On our end all we can really do is ensure that we're not checking grants if we don't need to--we wouldn't want to call `ListReferenceGrants()` for every plugin, but if it's a remote plugin, we know we need to check them.

File: internal/gatewayapi/references.go (hunk 7 above)
Comment (rainest) [reply internal/gatewayapi/references.go:79]: That'll essentially be part of https://github.com/Kong/kubernetes-ingress-controller/issues/6000

We plan to do away with the per-Kind cases entirely, and will unit test the forthcoming `client.Object`-based implementation. Not much worth writing them for the implementation we plan to discard in short order.

**Review by randmonkey — APPROVED** [review]
